Sources/Swift/HangTracker.swift

```diff
@@ -0,0 +1,216 @@
+            queue.async { [weak self] in
+                if self?.lateRunLoop.isEmpty ?? false {
+                    DispatchQueue.main.async {
+                        if self?.finishedRunLoop.isEmpty ?? false {
+                            self?.stop()
+                        }
+                    }
+                }
+            }
+        }
+    }
+    
+    // This queue is used to detect main thread hangs, they need to be detected on a background thread
+    // since the main thread is hanging.
+    private let queue: DispatchQueue
+    private let hangNotifyThreshold: TimeInterval
+    private let dateProvider: SentryCurrentDateProvider
+    private let createObserver: CreateObserverFunc<T>
+    private let addObserver: AddObserverFunc<T>
+    private let removeObserver: RemoveObserverFunc<T>
+    
+    // MARK: Main queue
+
+    private var observer: T?
+    private var finishedRunLoop = [UUID: (RunLoopIteration) -> Void]()
+    private var semaphore: DispatchSemaphore?
+    private var loopStartTime: TimeInterval?
+    
+    private func startIfNecessary() {
+        guard observer == nil else {
+            // Already running
+            return
+        }
+
+        let observer = createObserver(nil, CFRunLoopActivity.beforeWaiting.rawValue | CFRunLoopActivity.afterWaiting.rawValue, true, CFIndex(INT_MAX)) { [weak self] _, activity in
+            guard let self else { return }
+
+            let currentTime = dateProvider.systemUptime()
+            switch activity {
+            case .beforeWaiting:
+                semaphore?.signal()
+                if let loopStartTime {
+                    for handler in finishedRunLoop.values {
+                        handler(RunLoopIteration(startTime: loopStartTime, endTime: currentTime))
+                    }
+                }
+            case .afterWaiting:
+                let started = currentTime
+                loopStartTime = currentTime
+                let localSemaphore = DispatchSemaphore(value: 0)
+                semaphore = localSemaphore
+                queue.async { [weak self] in
+                    self?.waitForHang(semaphore: localSemaphore, started: started, isStarting: true)
+                }
+            default:
+                fatalError()
+            }
+        }
+        self.observer = observer
+        addObserver(CFRunLoopGetMain(), observer, .commonModes)
+    }
+    
+    private func stop() {
+        dispatchPrecondition(condition: .onQueue(.main))
+
+        guard let observer else {
+            return
+        }
+        removeObserver(CFRunLoopGetMain(), observer, .commonModes)
+        self.observer = nil
+    }
+    
+    // MARK: Background queue
+    
+    private var lateRunLoop = [UUID: (UUID, TimeInterval) -> Void]()
+    private var hangId = UUID()
+    
+    private func waitForHang(semaphore: DispatchSemaphore, started: TimeInterval, isStarting: Bool) {
+        dispatchPrecondition(condition: .onQueue(queue))
```

Comment:
`h`: We do not use `dispatchPrecondition` in the SDK because it causes runtime crashes. Please use a soft or dispatching to main